[WasmFS] Fix absolute path access under NODERAWFS

kleisauke opened the pull request:
Resolves: #24830.
Supersedes: #24833.

kleisauke committed:
[WasmFS] Fix absolute path access under NODERAWFS

Resolves: #24830.

kleisauke commented:
Marking as ready for review with a few caveats:
- `wasmfs.test_readdir_rawfs` appears to be failing.
- It also fails in the wasm-vips Windows CI when writing to `C:\Users\RUNNER~1\AppData\Local\Temp`, see:
  https://github.com/kleisauke/wasm-vips/actions/runs/23491851397/job/68365978350
  (this was workarounded in commit https://github.com/kleisauke/wasm-vips/commit/52f2d3cbea1574086a2dd05e8f66fc38e39314fd)
- Code size expectations need to be rebaselined.
- Some changes could be split into separate PRs.

tlively commented on system/lib/wasmfs/backends/node_backend.cpp:
```diff
@@ -296,6 +304,10 @@ class NodeBackend : public Backend {
     WASMFS_UNREACHABLE("TODO: implement NodeBackend::createSymlink");
   }
+
+  virtual bool requiresPathTraversal() override {
```
How about calling this `usesRawPaths`? (And flipping the result, of course.) I think that's slightly easier to understand than `requiresPathTraversal`, though it's not perfect.

tlively replied:
That sounds good to me.

tlively reviewed (commented):
Thanks for splitting this up. It's much more focused and easier to review now. Mostly LGTM, with some bikeshedding and a correctness concern.

kleisauke commented on system/lib/wasmfs/paths.cpp:
```diff
@@ -66,12 +66,21 @@ ParsedParent doParseParent(std::string_view path,
+  // full path directly. In that case, skip WasmFS path parsing and pass the
+  // path through unchanged.
+  if (!root->getBackend()->requiresPathTraversal()) {
+    return {std::make_pair(std::move(curr), path)};
```
Indeed, this is going to break in such sitations as it implicitly assumes that the entire file system behaves like the root backend.

Commit e21bc63820ff7f89947102450879e78301381fd3 changes this to `curr->getBackend()` instead, which allows:
```console
$ tree / -L 2 --dirsfirst
/ (MEMFS)
├── dev
│   ├── null -> SpecialFiles::getNull()
│   ├── random -> SpecialFiles::getRandom()
│   ├── stderr -> SpecialFiles::getStderr()
│   ├── stdin -> SpecialFiles::getStdin()
│   ├── stdout -> SpecialFiles::getStdout()
│   └── urandom -> SpecialFiles::getURandom()
└── host (NODERAWFS)
```
(i.e. the Node filesystem is mounted under `/host`, while the `wasmfs_create_root_dir()` hook is not overriden)

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (2) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_cxx_wasmfs.json: 179507 => 179652 [+145 bytes / +0.08%]
codesize/test_codesize_files_wasmfs.json: 63720 => 63900 [+180 bytes / +0.28%]

Average change: +0.18% (+0.08% - +0.28%)
```

kleisauke committed:
Merge branch 'main' into wasmfs-noderawfs-abspath

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (2) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_cxx_wasmfs.json: 179507 => 179652 [+145 bytes / +0.08%]
codesize/test_codesize_files_wasmfs.json: 63764 => 63944 [+180 bytes / +0.28%]

Average change: +0.18% (+0.08% - +0.28%)
```

kleisauke commented:
Looking at the `chdir()` and `wasmfs_root_override.cpp‎` workaround in PR https://github.com/dotnet/runtime/pull/130141, I think .NET could also benefit from this.

/cc @pavelsavara FYI.

pavelsavara commented:
- when using NODERAWFS, could also default virtual cwd match the host cwd ?
- we also have hack for running on windows. Especially `D:\runtime` instead of `C:\runtime` was not working well. This is with legacy VFS too.

https://github.com/dotnet/runtime/blob/aa7a6fdc5db9aecfeba30881a9dae2fc0cce0f59/src/coreclr/hosts/corerun/wasm/libCorerun.js#L262-L270

Sorry if that's off topic here.

kleisauke commented:
@pavelsavara No worries, I'm happy to answer these questions here.

> * when using NODERAWFS, could also default virtual cwd match the host cwd ?

The current working directory will be whatever the OS working directory was when you started your Node.js program. See for example:

<details>
  <summary>Details</summary>

`test.c`:
```c
#include <stdio.h>
#include <dirent.h>

int main(void) {
    DIR *d = opendir(".");

    struct dirent *ent;
    while ((ent = readdir(d)) != NULL)
        puts(ent->d_name);

    closedir(d);
    return 0;
}
```

```console
$ emcc test.c -sWASMFS -sNODERAWFS
$ mkdir work
$ cd work
$ touch NEW_FILE
$ node ../a.out.js
.
..
NEW_FILE
```
</details>

Though, I just noticed that `getcwd()` is not aware of this under `WASMFS` mode, so that still needs to be fixed. See:

<details>
  <summary>Details</summary>

`test.c`:
```c
#include <stdio.h>
#include <unistd.h>
#include <limits.h>

int main(void) {
    char cwd[PATH_MAX];
    puts(getcwd(cwd, sizeof(cwd));
    return 0;
}
```

```console
$ echo $PWD
/home/kleisauke/emscripten
$ emcc test.c -sWASMFS -sNODERAWFS
$ node a.out.js
/
$ emcc test.c -sNODERAWFS
$ node a.out.js
/home/kleisauke/emscripten
```
</details>

> * we also have hack for running on windows. Especially `D:\runtime` instead of `C:\runtime` was not working well. This is with legacy VFS too.

I think this hack can be removed after PR #17849 landed.

kleisauke committed:
Merge branch 'main' into wasmfs-noderawfs-abspath

kleisauke committed:
Automatic rebaseline of codesize expectations. NFC

This is an automatic change generated by tools/maint/rebaseline_tests.py.

The following (1) test expectation files were updated by
running the tests with `--rebaseline`:

```
codesize/test_codesize_cxx_wasmfs.json: 179507 => 179652 [+145 bytes / +0.08%]

Average change: +0.08% (+0.08% - +0.08%)
```

kleisauke committed:
Simplify
